src/main/java/com/microsoft/azure/relay/ClientWebSocket.java

```diff
@@ -173,36 +174,41 @@ boolean isOpen() {
 	 * @throws TimeoutException thrown when a complete message frame is not received within the timeout.
 	 */
 	CompletableFuture<ByteBuffer> readBinaryAsync(Duration timeout) {
-		
-		return CompletableFutureUtil.timedSupplyAsync(timeout, () -> {
-				LinkedList<byte[]> fragments = new LinkedList<byte[]>();
-				boolean receivedWholeMsg;
-				int messageSize = 0;
-				
-				do {
-					MessageFragment fragment = fragmentQueue.dequeueAsync().join();
-					if (fragment == null) {
-						break;
-					}				
-					messageSize += fragment.getBytes().length;
-					fragments.add(fragment.getBytes());
-					receivedWholeMsg = fragment.isEnd();
-				} 
-				while (!receivedWholeMsg);
-	
-				byte[] message = new byte[messageSize];
-				int offset = 0;
-				for (byte[] bytes : fragments) {
-					System.arraycopy(bytes, 0, message, offset, bytes.length);
-					offset += bytes.length;
+		LinkedList<byte[]> fragments = new LinkedList<byte[]>();
+		AtomicInteger messageSize = new AtomicInteger(0);
+		return readFragmentsAsync(fragments, messageSize, timeout).thenApply((nullResult) -> {
+			byte[] message = new byte[messageSize.get()];
+			int offset = 0;
+			for (byte[] bytes : fragments) {
+				System.arraycopy(bytes, 0, message, offset, bytes.length);
+				offset += bytes.length;
+			}
+			RelayLogger.logEvent("receivedBytes", this, String.valueOf(message.length));
+			return ByteBuffer.wrap(message);
+		});
+	}
+
+	private CompletableFuture<Void> readFragmentsAsync(LinkedList<byte[]> fragments, AtomicInteger messageSize, Duration timeout) {
+		TimeoutHelper timeoutHelper = (timeout != null) ? new TimeoutHelper(timeout, true) : null;
```

Comment:
>timeoutHelper [](start = 16, length = 13)

You could pass a single TimeoutHelper just like you pass the LinkedList and AtomicInteger. Depends if the timeout applies for the overall operation or for each fragment. I personally think it should be for the overall receive so a single TimeoutHelper might work well.